enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+
+Formalizing this guidance into a type of template or specification will allow
+developers and administrators to speak a common language and have a common
+understanding, even if the concepts seem initially unituitive.
+
+### Risks and Mitigations
+
+People will make wrong assumptions about alerts and how to handle them.  This
+is not unique to alerts, people often believe certain components should behave
+in a particular way in which they do not.  This is an artifact of poor
+documentation that leaves too much to the imagination.
+
+We can work around this problem with clear and concise documentation.
+
+## Design Details
+
+### Critical Alerts
+
+TL/DR:  Fix this now or cluster dies.
+
+Timeline:  ~5 minutes.
+
+Critical Alerts should be reserved for situations that jeopardize the health
+if not immediately addressed or if the cluster is already in a critical state.
+
+Some critical states are:
+* loss or impending loss of etcd-quorum.
+* inability to route application traffic externally or internally
+(dataplane disruption)
+* inability to start any new pod other than capacity.  EG, if the API server
+were to restart or need to be rescheduled, it would not able to start on another
+host
+
+In otherwords, critical alerts are something that require someone to get out
+of bed in the middle of the night and fix something **right now** or they will
+be faced with a disaster.
+
+An example of something that is NOT a critical alert:
+* MCO and/or related components are completely dead/crash looping.
+
+Even though the above is quite significant, there is no risk of immediate loss
+of the cluster.
+
+You might be thinking to yourself: "But my component is **super** important, and
+if it's not working, the user can't do X, which user definitely will want."  All
+that is probably true, but that doesn't make your alert "critical" it just makes
+it worthy of an alert.
+
+### Warning Alerts
+
+TL/DR: fix this soon, or some things won't work and upgrades will probably be
+blocked.
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs
+to the warning level or lower.
+
+Most alerts probably fit in this category.  This level should represent degraded
+or unavailable services that do not jeopardize the immediate life of cluster.
+
+Timeline:  ~60 minutes
+
+60 minutes?  That seems high!  That's because it is high.  We want to reduce
+the noise.  We've done a lot to make clusters and operators auto-heal
+themselves.  The whole idea is that if a condition has persisted for more than
+60 minutes, it's unlikely to be resolved without intervention any time later.
+
+#### Example 1: All machine-config-daemons are crashlooping.
+
+That seems significant
+and it probably is, but nobody has to get out of bed to fix this **right now**.
+But, what if a machine dies, I won't be able to get a replacement?  Yeah, that
+is probably true, but how often does that happen?  Also, that's an entirely
```

Comment:
"doesn't happen often" is not an argument for "not critical".  The argument for this not being critical is that there is no risk of data-loss or customer impact inherent in a machine dying, or failing to provision a new one, or rolling out new secrets to nodes, or all the other things that the machine-config does.  The critical impact would be "my customer-facing workload is failing to schedule because I have no capacity" or "I am on the edge of etcd-quorum loss because I cannot revive my third control-plane node".  Those are critical.  Maybe they are due to the machine-config stack being sad, so it's good to have warning-level machine-config alerts.  But I don't think that they need to be critical-level machine-config alerts.  But "happens rarely" is an argument for "even if the severity is wrong, fixing it is not all that important", not an argument for what the severity should be.

Reply:
> "my customer-facing workload is failing to schedule because I have no capacity"

No, we can't own this.  Users should have their own monitoring for apps that fail to schedule.  There's all sorts of capacity user stories and users need to use careful consideration in this area.

The machine-api does not replace failed instances anyway, MHC does that.

I disagree that frequency is not a component of severity.  If clusters routinely lost 1 instance an hour (in some hypothetical reality... there's a joke to be made about a cloud provider in here lol) just as a matter of course, then having a function machine-api would be critical.

Reply:
> Users should have their own monitoring for apps that fail to schedule.

User workloads should have "failed to schedule" monitoring so they can stay under the current caps and quotas.  But cluster admins should have "limited capacity is impacting user workloads" monitoring so they can decide when they want to grow quota.  This is the same as the autoscaler, which definitely cares about user workload capacity today, despite being managed by cluster admins.  The alerts would fill the manually-scaled use case, and also cover the "autoscaler is maxed / dead" use case.

> The machine-api does not replace failed instances anyway, MHC does that.

I thought the MHC killed dead/degraded machines/nodes, but that the machine API then provisioned the replacement Machine to get the MachineSet back up to it's `spec.replicas`.  If you have a MHC on a machine/node that is not part of a MachineSet, is a replacement created?  I'd have expected no replacement to be created.

> If clusters routinely lost 1 instance an hour (in some hypothetical reality... then having a function machine-api would be critical.

But it would not be a `critical` alert.  Losing your customer-facing workloads or living on the edge of etcd quorum loss would be a `critical` alert.  And then there would be `machine API is dead`, `high machine death rate`, and `unable to schedule many pods due to limited overall cluster capacity` `warning` alerts to help point the responding admin at likely root causes.

Reply:
> "limited capacity is impacting user workloads" monitoring so they can decide when they want to grow quota

This assumes having pending pods is a problem.  Some clusters might schedule very large batches of jobs, having limited capacity is not a problem for them, they just want to chew through the jobs.  Like I said, there are all sorts of capacity user stories.  A cluster can run indefinitely with 0 capacity.

> If you have a MHC on a machine/node

I can't recall MHC specifics, but it's not intended to delete machines that aren't part of a machineset today.  The machineset creates the replacement machine, but in effect, you can paraphrase that the MHC replaces the failed machine.